x-pack/filebeat/input/gcs/scheduler.go

```diff
@@ -0,0 +1,253 @@
+		// work distribution remains efficient
+		pager = s.fetchObjectPager(ctxWithTimeout, availableWorkers)
+		err := lmtr.scheduleOnce(ctxWithTimeout, pager, s)
+		if err != nil {
+			return err
+		}
+
+		err = timed.Wait(ctx, s.src.PollInterval)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// acquire gets an available worker thread.
+func (l *limiter) acquire() {
+	l.wg.Add(1)
+	l.limit <- struct{}{}
+}
+
+// release puts pack a worker thread.
+func (l *limiter) release() {
+	<-l.limit
+	l.wg.Done()
+}
+
+func (l *limiter) scheduleOnce(ctx context.Context, pager *iterator.Pager, s *gcsInputScheduler) error {
+	for {
+		var objects []*storage.ObjectAttrs
+		nextPageToken, err := pager.NextPage(&objects)
+		if err != nil {
+			return err
+		}
+
+		jobs := s.createJobs(objects, s.log)
+
+		// If previous checkpoint was saved then look up starting point for new jobs
+		if s.state.Checkpoint().LatestEntryTime != nil {
+			jobs = s.moveToLastSeenJob(jobs)
+			if len(s.state.Checkpoint().FailedJobs) > 0 {
+				jobs = s.addFailedJobs(ctx, jobs)
+			}
+		}
+
+		// distributes jobs among workers with the help of a limiter
+		for i, job := range jobs {
+			id := fetchJobID(i, s.src.BucketName, job.Name())
+			job := job
+			l.acquire()
+			go func() {
+				defer l.release()
+				job.Do(l.ctx, id)
+			}()
+		}
+
+		if nextPageToken == "" {
+			break
+		}
+	}
+	return nil
+}
+
+func (s *gcsInputScheduler) createJobs(objects []*storage.ObjectAttrs, log *logp.Logger) []job.Job {
+	var jobs []job.Job
+
+	for _, obj := range objects {
+		// check required to ignore directories & sub folders, since there is no inbuilt option to
+		// do so. In gcs all the directories are emulated and represented by a prefix, we can
+		// define specific prefix's & delimiters to ignore known directories but there is no generic
+		// way to do so.
+		file := strings.Split(obj.Name, "/")
+		if len(file) > 1 && file[len(file)-1] == "" {
+			continue
+		}
+
+		objectURI := "gs://" + s.src.BucketName + "/" + obj.Name
+		job := job.NewGcsInputJob(s.bucket, obj, objectURI, s.state, s.src, s.publisher, log, false)
+		jobs = append(jobs, job)
+	}
+
+	return jobs
+}
+
+// fetchObjectPager fetches the page handler for objects, given a batch size.
+// [NOTE] : There are no api's / sdk functions that list blobs via timestamp/latest entry, it's always lexicographical order
+func (s *gcsInputScheduler) fetchObjectPager(ctx context.Context, pageSize int) *iterator.Pager {
+	bktIt := s.bucket.Objects(ctx, &storage.Query{})
+	pager := iterator.NewPager(bktIt, pageSize, "")
+
+	return pager
+}
+
+// moveToLastSeenJob, moves to the latest job position past the last seen job
+// Jobs are stored in lexicographical order always , hence the latest position can be found either on the basis of job name or timestamp
+func (s *gcsInputScheduler) moveToLastSeenJob(jobs []job.Job) []job.Job {
+	var latestJobs []job.Job
+	jobsToReturn := make([]job.Job, 0)
```

Comment:
```suggestion
	var jobsToReturn []job.Job
```

Modulo name and type change.